### DISCUSSION TITLE
How to fix a bad input in an event?

### DISCUSSION THREAD

**Discussion by Punksolid** [spatie/laravel-event-sourcing#452] [Q&A]
Hi everyone,

I've been working with Laravel Event Sourcing and have encountered a challenge that I'd appreciate some advice

I initially implemented Events and Projectors for updating account balances. This setup has worked well for several months, updating balances several times a day. My setup includes:

    A projector for updating account balances.
    Another projector for updating account balance historical points.
    A new projector I'm working on for an overall user balance historical point.

However, I've run into an issue where some events registered unusually high balances due to a bug. I've fixed the bug, but it left some historical data incorrect. When replaying events for all projectors, or only new ones, it results in inaccurate historical points.

I've read about Compensation Events and Retroactive Events as potential solutions. But when I tried implementing these, I couldn't properly trigger a new event/projector inside my normal projectors. It also seemed like bad practice to add a conditional in every affected projector.  I'm concerned about the sustainability of this approach as future errors might arise.

Could anyone advise on the recommended process to handle such a scenario in Laravel Event Sourcing? Is there a more efficient way to correct historical data without compromising the integrity of the event log?

```php
class AccountBalanceProjector extends Projector
{
    public function onBalanceUpdated(BalanceUpdated $event): void
    {
        $account = Account::findOrFail($event->accountId);

        if ($event->createdAt()->between('2023-10-20 00:00:00', '2023-11-09 23:59:59') &&  $event->balance > 35000000) {
            // if account balance has more than a million
            event(
                new AccountBalanceCompensationEvent( //the event is dispatched but the projector related is not
                $event->accountId,
                1500000, // hardcoded as example
                    'automatic'
                )
            );
        } else {
            $account->balance = $event->balance;
            $account->updated_at = $event->createdAt();
            $account->save();
        }
    }
}
```
For now I'm just doing a skip of the actions if it encounter a bad scenario but I feel is not the best approach which I consider safe because I will keep all the events untouched.

Thank you in advance for your insights!

**Comment by sebastiandedeyne** [comment]
This is technically bad advice, but I don't shy away from writing a migration that fixes old events if they were caused by a bug in the past. I don't want a but that lasted a few days to be something I have to deal with in code for the next 10 years. Migrating old events is risky but gets rid of the problem for good.

**Reply by Punksolid** [reply]
@sebastiandedeyne thanks for your advice, yeah that was also my first thought, in fact the project is just some months old but I wanted to do it the "right way". I will keep it in mind because currently I consider the hardcoded conditional to skip an event almost as bad as doing a fix on the original events.
